crypto_stream_chacha20.js

```diff
@@ -0,0 +1,214 @@
+    'k should be crypto_stream_chacha20_ietf_KEYBYTES')
+
+  var xor = new Chacha20(n, k)
+  xor.update(c, m)
+  xor.final()
+}
+
+exports.crypto_stream_chacha20_ietf_xor_ic = function (c, m, mlen, n, ic, k) {
+  assert(n.byteLength === exports.crypto_stream_chacha20_ietf_NONCEBYTES,
+    'n should be crypto_stream_chacha20_ietf_NONCEBYTES')
+  assert(k.byteLength === exports.crypto_stream_chacha20_ietf_KEYBYTES,
+    'k should be crypto_stream_chacha20_ietf_KEYBYTES')
+
+  var xor = new Chacha20(n, k, ic)
+  xor.update(c, m)
+  xor.final()
+}
+
+exports.crypto_stream_chacha20_ietf_xor_instance = function (n, k) {
+  assert(n.byteLength === exports.crypto_stream_chacha20_ietf_NONCEBYTES,
+    'n should be crypto_stream_chacha20_ietf_NONCEBYTES')
+  assert(k.byteLength === exports.crypto_stream_chacha20_ietf_KEYBYTES,
+    'k should be crypto_stream_chacha20_ietf_KEYBYTES')
+
+  return new Chacha20(n, k)
+}
+
+function Chacha20 (n, k, counter) {
+  assert(k.byteLength === exports.crypto_stream_chacha20_ietf_KEYBYTES)
+  assert(n.byteLength === exports.crypto_stream_chacha20_NONCEBYTES ||
+    n.byteLength === exports.crypto_stream_chacha20_ietf_NONCEBYTES)
+
+  if (!counter) counter = 0
+  assert(counter < Number.MAX_SAFE_INTEGER)
+
+  this.finalized = false
+  this.pos = 0
+  this.state = new Uint32Array(16)
+
+  for (let i = 0; i < 4; i++) this.state[i] = constant[i]
+  for (let i = 0; i < 8; i++) this.state[4 + i] = k.readUInt32LE(4 * i)
+
+  this.state[12] = counter & 0xffffffff
+
+  if (n.byteLength === 8) {
+    this.state[13] = counter >> 32
+    this.state[14] = n.readUInt32LE(0)
+    this.state[15] = n.readUInt32LE(4)
+  } else {
+    this.state[13] = n.readUInt32LE(0)
+    this.state[14] = n.readUInt32LE(4)
+    this.state[15] = n.readUInt32LE(8)
+  }
+
+  return this
+}
+
+Chacha20.prototype.update = function (output, input) {
+  assert(!this.finalized, 'cipher finalized.')
+  assert(output.byteLength >= input.byteLength,
+    'output cannot be shorter than input.')
+
+  var len = input.length
+  var offset = this.pos % 64
+  this.pos += len
+
+  // input position
+  var j = 0
+
+  var keyStream = chacha20_block(this.state)
+
+  // try to finsih the current block
+  while (offset > 0 && len > 0) {
+    output[j] = input[j++] ^ keyStream[offset]
+    offset = (offset + 1) & 0x3f
+    if (!offset) this.state[12]++
+    len--
+  }
+
+  // encrypt rest block at a time
+  while (len > 0) {
+    keyStream = chacha20_block(this.state)
+
+    // less than a full block remaining
+    if (len < 64) {
+      for (let i = 0; i < len; i++) {
+        output[j] = input[j++] ^ keyStream[offset++]
+        offset &= 0x3f
+      }
+
+      return
+    }
+
+    for (; offset < 64;) {
+      output[j] = input[j++] ^ keyStream[offset++]
+    }
+
+    this.state[12]++
+    offset = 0
+    len -= 64
+  }
+}
+
+Chacha20.prototype.final = function () {
+  this.finalized = true
+}
+
+module.exports.keystream = function (output, key, nonce, counter) {
```

Comment:
is this a sodium api?

Reply:
old code, this api is not in libsodium. can be accessed using `crypto_stream_chacha20_xor_ic` methods with empty plaintext instead